front/src/modules/simulationResult/components/SpaceTimeChart/useLazyProjectTrains.ts

```diff
@@ -0,0 +1,158 @@
+import { useEffect, useState, type Dispatch, type SetStateAction, useMemo } from 'react';
+
+import { keyBy } from 'lodash';
+import { useSelector } from 'react-redux';
+
+import upsertNewProjectedTrains from 'applications/operationalStudies/helpers/upsetNewProjectedTrains';
+import type { TrainSpaceTimeData } from 'applications/operationalStudies/types';
+import {
+  osrdEditoastApi,
+  type PathfindingResultSuccess,
+  type TrainScheduleResult,
+} from 'common/api/osrdEditoastApi';
+import { useOsrdConfSelectors } from 'common/osrdContext';
+import type { Dictionary } from 'utils/types';
+
+import { getTrainIdsPackage } from './utils';
+
+const BATCH_SIZE = 10;
+
+type useLazyLoadTrainsProp = {
+  infraId?: number;
+  trainIdsToProject: number[];
+  path?: PathfindingResultSuccess;
+  trainSchedules?: TrainScheduleResult[];
+  moreTrainsToCome?: boolean;
+  setTrainIdsToProject: Dispatch<SetStateAction<number[]>>;
+};
+
+/**
+ * This hook gradually projects trains of the timetable.
+ *
+ * It projects the trains by batch of 10 on the path.
+ * This optimizes the performance of the application and allow us to display the
+ * trains on the space time chart as soon as they are ready.
+ */
+const useLazyProjectTrains = ({
+  infraId,
+  trainIdsToProject,
+  path,
+  trainSchedules,
+  moreTrainsToCome = false,
+  setTrainIdsToProject,
+}: useLazyLoadTrainsProp) => {
+  const { getElectricalProfileSetId } = useOsrdConfSelectors();
+  const electricalProfileSetId = useSelector(getElectricalProfileSetId);
+
+  const [projectedTrainsById, setProjectedTrainsById] = useState<Dictionary<TrainSpaceTimeData>>(
+    {}
+  );
+  const projectedTrains = useMemo(() => Object.values(projectedTrainsById), [projectedTrainsById]);
+
+  const [nextPackageToProjectIndex, setNextPackageToProjectIndex] = useState(0);
+  const [readyToProjectNextPackage, setReadyToProjectNextPackage] = useState(true);
+
+  const [postV2TrainScheduleProjectPath] =
+    osrdEditoastApi.endpoints.postV2TrainScheduleProjectPath.useLazyQuery();
+
+  const trainSchedulesById = useMemo(() => keyBy(trainSchedules, 'id'), [trainSchedules]);
+
+  useEffect(() => {
+    // reset the state when all the trains have been projected
+    if (
+      !moreTrainsToCome &&
+      readyToProjectNextPackage &&
+      nextPackageToProjectIndex * 10 > trainIdsToProject.length
+    ) {
+      setNextPackageToProjectIndex(0);
+      setTrainIdsToProject([]);
+    }
+  }, [moreTrainsToCome, readyToProjectNextPackage]);
```

Comment:
Do we really need to reset the state? Can't we leave `trainIdsToProject` as-is?

Reply:
Here is a suggestion based on what we started to work on together yesterday. It uses references to persist pieces of information across component re-renders (when it's unnecessary to use these as dependencies of `useEffect`), and a sequence number of the currently active projection to prevent multiple from running in parallel.

```ts
/**
 * This hook gradually projects trains of the timetable.
 *
 * It projects the trains by batch of 10 on the path.
 * This optimizes the performance of the application and allow us to display the
 * trains on the space time chart as soon as they are ready.
 */
const useLazyProjectTrains = ({
  infraId,
  trainIdsToProject,
  path,
  trainSchedules,
  moreTrainsToCome = false,
  setTrainIdsToProject,
}: useLazyLoadTrainsProp) => {
  const { getElectricalProfileSetId } = useOsrdConfSelectors();
  const electricalProfileSetId = useSelector(getElectricalProfileSetId);

  const [projectedTrainsById, setProjectedTrainsById] = useState<Dictionary<TrainSpaceTimeData>>(
    {}
  );
  const projectedTrains = useMemo(() => Object.values(projectedTrainsById), [projectedTrainsById]);

  const requestedProjectedTrainIds = useRef<Set<number>>(new Set());
  const projectionSeqNum = useRef<number>(0);

  const [postV2TrainScheduleProjectPath] =
    osrdEditoastApi.endpoints.postV2TrainScheduleProjectPath.useLazyQuery();

  const trainSchedulesById = useMemo(() => keyBy(trainSchedules, 'id'), [trainSchedules]);

  // gradually project the trains on the selected path
  useEffect(() => {
    const projectNextPackage = async (
      _pathfinding: PathfindingResultSuccess,
      packageToProject: number[]
    ) => {
      packageToProject.forEach((trainId) => requestedProjectedTrainIds.current.add(trainId));

      const { blocks, routes, track_section_ranges } = _pathfinding;
      const rawProjectedTrains = await postV2TrainScheduleProjectPath({
        projectPathForm: {
          infra_id: infraId!,
          ids: packageToProject,
          path: { blocks, routes, track_section_ranges },
          electrical_profile_set_id: electricalProfileSetId,
        },
      }).unwrap();

      setProjectedTrainsById((prevTrains) => {
        const newProjectedTrains = upsertNewProjectedTrains(
          prevTrains,
          rawProjectedTrains,
          trainSchedulesById
        );
        return newProjectedTrains;
      });
      // console.log('2. package', packageToProject);
    };

    const projectTrains = async (seqNum: number, path: PathfindingResultSuccess, _trainToProjectIds: number[]) => {
      const shouldProjectIds = _trainToProjectIds.filter(
        (trainId) => !requestedProjectedTrainIds.current.has(trainId)
      );

      for (let i = 0; i * BATCH_SIZE < shouldProjectIds.length; i += 1) {
        // If projection parameters have changed, bail out
        if (projectionSeqNum.current !== seqNum) break;

        const packageToProject = getTrainIdsPackage(i, shouldProjectIds, BATCH_SIZE);
        try {
          await projectNextPackage(path, packageToProject);
        } catch (e) {
          console.error('error', e);
        }
      }
    };

    if (infraId && path) {
      projectionSeqNum.current += 1;
      projectTrains(projectionSeqNum.current, path, trainIdsToProject);
    }
  }, [trainIdsToProject]);

  return {
    projectedTrains,
    setProjectedTrainsById,
  };
};
```

Reply:
Done, with the refacto of useLazyProjectTrains.
Thanks for the suggestion !